internal/cache/file/cache_handler_test.go

```diff
@@ -0,0 +1,438 @@
+
+	ExpectEq(nil, err)
+	jobStatus := cacheHandle.fileDownloadJob.GetStatus()
+	ExpectEq(downloader.INVALID, jobStatus.Name)
+	ExpectEq(false, isFileExist(chrT.downloadPath))
+}
+
+func (chrT *cacheHandlerTest) Test_InvalidateCache_WhenEntryNotInCache() {
+	minObject := chrT.getMinObject("object_1", "content of object_1")
+
+	err := chrT.cacheHandler.InvalidateCache(minObject, chrT.bucket)
+
+	ExpectEq(nil, err)
+}
+
+func (chrT *cacheHandlerTest) Test_InvalidateCache_ConcurrentSameFile() {
+	// Existing cacheHandle.
+	cacheHandle1 := chrT.getCacheHandleForSetupEntryInCache()
+	AssertEq(nil, cacheHandle1.validateCacheHandle())
+
+	wg := sync.WaitGroup{}
+
+	InvalidateCacheTestFun := func() {
+		defer wg.Done()
+		minObj := chrT.getMinObject("object_1", "content of object_1 ...")
+
+		err := chrT.cacheHandler.InvalidateCache(minObj, chrT.bucket)
+
+		AssertEq(nil, err)
+	}
+
+	// Start concurrent GetCacheHandle()
+	for i := 0; i < 5; i++ {
+		wg.Add(1)
+		go InvalidateCacheTestFun()
+	}
+	wg.Wait()
+}
+
+func (chrT *cacheHandlerTest) Test_InvalidateCache_ConcurrentDifferentFiles() {
+	// Existing cacheHandle.
+	cacheHandle1 := chrT.getCacheHandleForSetupEntryInCache()
+	AssertEq(nil, cacheHandle1.validateCacheHandle())
+	wg := sync.WaitGroup{}
+
+	InvalidateCacheTestFun := func(index int) {
+		defer wg.Done()
+		objName := "object" + strconv.Itoa(index)
+		objContent := "object content: " + strconv.Itoa(index)
+		minObj := chrT.getMinObject(objName, objContent)
+
+		err := chrT.cacheHandler.InvalidateCache(minObj, chrT.bucket)
+
+		AssertEq(nil, err)
+	}
+
+	// Start concurrent GetCacheHandle()
+	for i := 0; i < 5; i++ {
+		wg.Add(1)
+		go InvalidateCacheTestFun(i)
+	}
+	wg.Wait()
+}
+
+func (chrT *cacheHandlerTest) Test_InvalidateCacheAndGetHandle_Concurrent() {
+	// Existing cacheHandle.
+	cacheHandle1 := chrT.getCacheHandleForSetupEntryInCache()
+	AssertEq(nil, cacheHandle1.validateCacheHandle())
+	wg := sync.WaitGroup{}
+
+	InvalidateCacheTestFun := func(index int) {
+		defer wg.Done()
+		objName := "object" + strconv.Itoa(index)
+		objContent := "object content: " + strconv.Itoa(index)
+		minObj := chrT.getMinObject(objName, objContent)
+
+		err := chrT.cacheHandler.InvalidateCache(minObj, chrT.bucket)
+
+		AssertEq(nil, err)
+	}
+
+	getCacheHandleTestFun := func(index int) {
+		defer wg.Done()
+		objName := "object" + strconv.Itoa(index)
+		objContent := "object content: " + strconv.Itoa(index)
+		minObj := chrT.getMinObject(objName, objContent)
+
+		cacheHandle, err := chrT.cacheHandler.GetCacheHandle(minObj, chrT.bucket, 0)
+
+		AssertEq(nil, err)
+		AssertEq(nil, cacheHandle.validateCacheHandle())
+	}
+
+	// Start concurrent GetCacheHandle()
+	for i := 0; i < 5; i++ {
+		wg.Add(1)
+		go InvalidateCacheTestFun(i)
+		wg.Add(1)
+		go getCacheHandleTestFun(i)
+	}
+	wg.Wait()
+
+	jobStatus := cacheHandle1.fileDownloadJob.GetStatus()
+	ExpectEq(downloader.INVALID, jobStatus.Name)
+	ExpectEq(false, isFileExist(chrT.downloadPath))
```

Comment:
we can't say for sure that InvalidateCacheTestFun is called in the last right ?

Reply:
That's true, but eviction of existing fileInfo (before the concurrent test started) is sure.

Reply:
Yes, however, ExpectEq(false, isFileExist(chrT.downloadPath)) should not always be true if InvalidateCacheTestFun is not called in the last ?